Class/Warrior/Spells/ShieldSlam.cpp

```diff
@@ -0,0 +1,87 @@
+#include "ShieldSlam.h"
+
+#include <cmath>
+
+#include "Buff.h"
+#include "CastComplete.h"
+#include "CharacterStats.h"
+#include "CombatRoll.h"
+#include "CooldownControl.h"
+#include "Equipment.h"
+#include "HeroicStrike.h"
+#include "Item.h"
+#include "MainhandAttack.h"
+#include "OffhandAttack.h"
+#include "Random.h"
+#include "Warrior.h"
+#include "WarriorSpells.h"
+#include "Weapon.h"
+
+ShieldSlam::ShieldSlam(Warrior* pchar, WarriorSpells* spells, CooldownControl* cooldown_control) :
+    Spell("Shield Slam", "Assets/ability/Ability_warrior_sunder.png", pchar, cooldown_control, RestrictedByGcd::Yes, ResourceType::Rage, 20),
+    TalentRequirer(QVector<TalentRequirerInfo*> {new TalentRequirerInfo("Shield Slam", 1, DisabledAtZero::Yes)}),
+    warr(pchar),
+    spells(spells),
+    shield_slam_range(new Random(340, 362)),
```

Comment:
Yup, this was vestigial of my initial attempt to make this scale with player level. What's the plans for that? There's evidence of it throughout the codebase but it's clearly not close.

Reply:
The plans for non-60 sims are effectively tied to making sure all relevant spells have their appropriate spell ranks set for their available spells. If you check something like `MageSpells` which is newer than `WarriorSpells` you'll see most spells have all their ranks implemented.

What then follows is the spell-to-rotation linking that is done as part of the preparations for combat. This already takes into account the availability of the spell rank when it tries to link spells, and by passing the CLVL the max ranks are restricted accordingly.

Further, base stats are currently hardcoded to 60, which is something that has to be resolved. Then there are special combat rules for below level 10 characters, but those are omitted in these plans.

The meta-reason I want to be able to sim lower level chars is because this entire sim project spawned from wondering what the optimal setup for a level ~35 warrior would be while killing Earth Elementals in Badlands..